README.md
[![Coverage Status](https://coveralls.io/repos/github/Stratio/kafka-elasticsearch-sink/badge.svg?branch=master)](https://coveralls.io/github/Stratio/kafka-elasticsearch-sink?branch=master)

# kafka-elasticsearch-sink

kafka-elasticsearch-sink is a library for generate tasks in Kafka Connect for connecting to Elasticsearch.

 
## Requirements

This library requires Kafka 0.10.0 and Elasticsearch 2.0.2


## Using the library

The ElasticsearchSinkTask can be configured with the following configuration.

  ```
    connector.class=org.apache.kafka.connect.es.KafkaElasticSearchSinkConnector,
    topics=metadata
    action.type=insert,
    elasticsearch.cluster.name=dg-cluster
    elasticsearch.hosts=127.0.0.1:9300
    elasticsearch.index=dg-metadata
    elasticsearch.mapping.type=type-v0
    elasticsearch.bulk.size=100000
  ```

By default the library insert data from Kafka topics, in addiction is possible to create Kafka Connect tasks in order
 to make updates, deletes or upserts in the Elasticsearch indexes.
 
Is necessary add the jar plugin inside the connectors directory and this directory should be added in the classpath:

```
  mkdir ${KAFKA_PATH}/libs_connect
  cp -r target/kafka-elasticsearch-sink-0.0.1-SNAPSHOT-jar-with-dependencies.jar ${KAFKA_PATH}/libs_connect/
  export CLASSPATH=${KAFKA_PATH}/libs_connect/*
```
  
If the user want make persistent this change can add this export in the file ".bashrc"
 

### Build

    `mvn clean package`


# License #

Licensed to STRATIO (C) under one or more contributor license agreements.
See the NOTICE file distributed with this work for additional information
regarding copyright ownership.  The STRATIO (C) licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.


CHANGELOG.md
# Changelog

## 0.1.0 (upcoming)

* Sink to save Kafka stream data into Elasticsearch


docker/bootstrap.sh
#!/usr/bin/env bash

set -o nounset \
    -o verbose \
    -o xtrace

[ -f /etc/confluent/docker/apply-mesos-overrides ] && \
	. /etc/confluent/docker/apply-mesos-overrides

set -o errexit

echo "===> ENV Variables ..."
env | sort

echo "===> User"
id

echo "===> Configuring ..."
/etc/confluent/docker/configure

if [[ ! -v CONNECT_TASK_NAME ]]; then
   CONNECT_TASK_NAME=kafka-connect-task-name
fi

if [[ ! -v KAFKA_TOPICS ]]; then
   KAFKA_TOPICS=topictest
fi

if [[ ! -v KAFKA_TOPIC_REPLICATION_FACTOR ]]; then
   KAFKA_TOPIC_REPLICATION_FACTOR=1
fi

if [[ ! -v KAFKA_TOPIC_PARTITIONS ]]; then
   KAFKA_TOPIC_PARTITIONS=1
fi

if [[ ! -v KAFKA_TOPICS ]]; then
   KAFKA_TOPICS=topictest
fi

if [[ ! -v ELASTICSEARCH_HOSTS ]]; then
   ELASTICSEARCH_HOSTS=localhost:9300
fi

if [[ ! -v ELASTICSEARCH_CLUSTER_NAME ]]; then
   ELASTICSEARCH_CLUSTER_NAME=elasticsearch
fi

if [[ ! -v ELASTICSEARCH_INDEX ]]; then
   ELASTICSEARCH_INDEX=connect-index
fi

if [[ ! -v ELASTICSEARCH_MAPPING_TYPE ]]; then
   ELASTICSEARCH_MAPPING_TYPE=mapping-v1
fi

if [[ ! -v ELASTICSEARCH_BULK_SIZE ]]; then
   ELASTICSEARCH_BULK_SIZE=250
fi

if [[ ! -v ELASTICSEARCH_ID_FIELD ]]; then
   ELASTICSEARCH_ID_FIELD=id
fi

if [[ ! -v ELASTICSEARCH_ACTION_TYPE ]]; then
   ELASTICSEARCH_ACTION_TYPE=insert
fi

if [[ ! -v CONNECT_CONVERTER_ENABLE_SCHEMAS ]]; then
   CONNECT_CONVERTER_ENABLE_SCHEMAS=false
fi
if [[ ! -v CONNECT_INTERNAL_CONVERTER_ENABLE_SCHEMAS ]]; then
   CONNECT_INTERNAL_CONVERTER_ENABLE_SCHEMAS=false
fi

CONNECT_CONF_FILE=/etc/kafka-connect/kafka-connect.properties

if [[ "${CONNECT_CONVERTER_ENABLE_SCHEMAS}" == "false" ]]; then
  sed -i "s|key.converter.schemas.enable=true*|key.converter.schemas.enable=false|" $CONNECT_CONF_FILE
  sed -i "s|value.converter.schemas.enable=true*|value.converter.schemas.enable=false|" $CONNECT_CONF_FILE
  echo "key.converter.schemas.enable=false" >> $CONNECT_CONF_FILE
  echo "value.converter.schemas.enable=false" >> $CONNECT_CONF_FILE
fi

if [[ "${CONNECT_INTERNAL_CONVERTER_ENABLE_SCHEMAS}" == "false" ]]; then
  sed -i "s|internal.key.converter.schemas.enable=true*|internal.key.converter.schemas.enable=false|" $CONNECT_CONF_FILE
  sed -i "s|internal.value.converter.schemas.enable=true*|internal.value.converter.schemas.enable=false|" $CONNECT_CONF_FILE
  echo "internal.key.converter.schemas.enable=false" >> $CONNECT_CONF_FILE
  echo "internal.value.converter.schemas.enable=false" >> $CONNECT_CONF_FILE
fi

echo "===> Running preflight checks ... "
/etc/confluent/docker/ensure

echo "===> Creating topics in Kafka"
topics=$(echo $KAFKA_TOPICS | tr "," "\n")
for topic in $topics
do
   echo "===> Creating topic $topic in Kafka"
   /usr/bin/kafka-topics --create --zookeeper $CONNECT_ZOOKEEPER_CONNECT --replication-factor $KAFKA_TOPIC_REPLICATION_FACTOR --partitions $KAFKA_TOPIC_PARTITIONS --topic $topic || echo "The topic $topic exists"
done

echo "===> Launching Kafka Connect ... "
exec /etc/confluent/docker/launch &

echo "===> Waiting for TCP connection to Kafka Connect ... "
while ! nc -w 1 127.0.0.1 8083
do
  sleep 1s
done
echo -n "done!"
echo "===> Kafka Connect Up"

echo "===> Sending tasks to Kafka Connect ..."
TASK_FILE=/etc/kafka-connect-tasks/elasticsearch-task.json

echo "===> Applying configuration to ${TASK_FILE}"
sed -i "s|\"name\":\"kafka.*|\"name\":\"${CONNECT_TASK_NAME}\",|" $TASK_FILE
sed -i "s|\"topics.*|\"topics\":\"${KAFKA_TOPICS}\",|" $TASK_FILE
sed -i "s|\"action.type.*|\"action.type\":\"${ELASTICSEARCH_ACTION_TYPE}\",|" $TASK_FILE
sed -i "s|\"elasticsearch.hosts.*|\"elasticsearch.hosts\":\"${ELASTICSEARCH_HOSTS}\",|" $TASK_FILE
sed -i "s|\"elasticsearch.cluster.name.*|\"elasticsearch.cluster.name\":\"${ELASTICSEARCH_CLUSTER_NAME}\",|" $TASK_FILE
sed -i "s|\"elasticsearch.index.*|\"elasticsearch.index\":\"${ELASTICSEARCH_INDEX}\",|" $TASK_FILE
sed -i "s|\"elasticsearch.mapping.type.*|\"elasticsearch.mapping.type\":\"${ELASTICSEARCH_MAPPING_TYPE}\",|" $TASK_FILE
sed -i "s|\"elasticsearch.id.field.*|\"elasticsearch.id.field\":\"${ELASTICSEARCH_ID_FIELD}\",|" $TASK_FILE
sed -i "s|\"elasticsearch.bulk.size.*|\"elasticsearch.bulk.size\":\"${ELASTICSEARCH_BULK_SIZE}\"|" $TASK_FILE

echo "===> Sending task to Kafka Connect ... "
curl -H "Content-Type: application/json; charset=UTF-8" -X POST http://localhost:8083/connectors -d @${TASK_FILE}

tail -f /dev/null


docker/docker-entrypoint.sh
#!/usr/bin/env bash

set -o nounset \
    -o verbose \
    -o xtrace

[ -f /etc/confluent/docker/apply-mesos-overrides ] && \
	. /etc/confluent/docker/apply-mesos-overrides

set -o errexit

echo "===> ENV Variables ..."
env | sort

echo "===> User"
id

echo "===> Configuring ..."
/etc/confluent/docker/configure

if [[ ! -v CONNECT_TASK_NAME ]]; then
   CONNECT_TASK_NAME=kafka-connect-task-name
fi

if [[ ! -v KAFKA_TOPICS ]]; then
   KAFKA_TOPICS=topictest
fi

if [[ ! -v ELASTICSEARCH_HOSTS ]]; then
   ELASTICSEARCH_HOSTS=localhost:9300
fi

if [[ ! -v ELASTICSEARCH_CLUSTER_NAME ]]; then
   ELASTICSEARCH_CLUSTER_NAME=elasticsearch
fi

if [[ ! -v ELASTICSEARCH_INDEX ]]; then
   ELASTICSEARCH_INDEX=connect-index
fi

if [[ ! -v ELASTICSEARCH_DOCUMENT_NAME ]]; then
   ELASTICSEARCH_DOCUMENT_NAME=mapping-v1
fi

if [[ ! -v ELASTICSEARCH_BULK_SIZE ]]; then
   ELASTICSEARCH_BULK_SIZE=250
fi

if [[ ! -v CONNECT_CONVERTER_ENABLE_SCHEMAS ]]; then
   CONNECT_CONVERTER_ENABLE_SCHEMAS=false
fi
if [[ ! -v CONNECT_INTERNAL_CONVERTER_ENABLE_SCHEMAS ]]; then
   CONNECT_INTERNAL_CONVERTER_ENABLE_SCHEMAS=false
fi

CONNECT_CONF_FILE=/etc/kafka-connect/kafka-connect.properties

if [[ "${CONNECT_CONVERTER_ENABLE_SCHEMAS}" == "false" ]]; then
  sed -i "s|key.converter.schemas.enable=true*|key.converter.schemas.enable=false|" $CONNECT_CONF_FILE
  sed -i "s|value.converter.schemas.enable=true*|value.converter.schemas.enable=false|" $CONNECT_CONF_FILE
  echo "key.converter.schemas.enable=false" >> $CONNECT_CONF_FILE
  echo "value.converter.schemas.enable=false" >> $CONNECT_CONF_FILE
fi

if [[ "${CONNECT_INTERNAL_CONVERTER_ENABLE_SCHEMAS}" == "false" ]]; then
  sed -i "s|internal.key.converter.schemas.enable=true*|internal.key.converter.schemas.enable=false|" $CONNECT_CONF_FILE
  sed -i "s|internal.value.converter.schemas.enable=true*|internal.value.converter.schemas.enable=false|" $CONNECT_CONF_FILE
  echo "internal.key.converter.schemas.enable=false" >> $CONNECT_CONF_FILE
  echo "internal.value.converter.schemas.enable=false" >> $CONNECT_CONF_FILE
fi

echo "===> Running preflight checks ... "
/etc/confluent/docker/ensure

echo "===> Launching Kafka Connect ... "
exec /etc/confluent/docker/launch &

echo "===> Waiting for TCP connection to Kafka Connect ... "
while ! nc -w 1 127.0.0.1 8083
do
  sleep 1s
done
echo -n "done!"
echo "===> Kafka Connect Up"

echo "===> Sending tasks to Kafka Connect ..."

TASK_FILE=/etc/kafka-connect-tasks/elasticsearch-task.json

echo "===> Applying configuration to ${TASK_FILE}"

sed -i "s|\"name\":\"kafka.*|\"name\":\"${CONNECT_TASK_NAME}\",|" $TASK_FILE
sed -i "s|\"topics.*|\"topics\":\"${KAFKA_TOPICS}\",|" $TASK_FILE
sed -i "s|\"elasticsearch.hosts.*|\"elasticsearch.hosts\":\"${ELASTICSEARCH_HOSTS}\",|" $TASK_FILE
sed -i "s|\"elasticsearch.cluster.name.*|\"elasticsearch.cluster.name\":\"${ELASTICSEARCH_CLUSTER_NAME}\",|" $TASK_FILE
sed -i "s|\"elasticsearch.index.*|\"elasticsearch.index\":\"${ELASTICSEARCH_INDEX}\",|" $TASK_FILE
sed -i "s|\"elasticsearch.document.name.*|\"elasticsearch.document.name\":\"${ELASTICSEARCH_DOCUMENT_NAME}\",|" $TASK_FILE
sed -i "s|\"elasticsearch.bulk.size.*|\"elasticsearch.bulk.size\":\"${ELASTICSEARCH_BULK_SIZE}\"|" $TASK_FILE

echo "===> Sending task to Kafka Connect ... "
curl -H "Content-Type: application/json; charset=UTF-8" -X POST http://localhost:8083/connectors -d @${TASK_FILE}

tail -f /dev/null


src/main/java/org/apache/kafka/connect/es/KafkaElasticSearchSinkTask.java
/*
 * Copyright (C) 2015 Stratio (http://stratio.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.kafka.connect.es;

import org.apache.kafka.clients.consumer.OffsetAndMetadata;
import org.apache.kafka.common.TopicPartition;
import org.apache.kafka.common.config.ConfigException;
import org.apache.kafka.connect.errors.ConnectException;
import org.apache.kafka.connect.errors.RetriableException;
import org.apache.kafka.connect.sink.SinkRecord;
import org.apache.kafka.connect.sink.SinkTask;
import org.elasticsearch.action.bulk.BulkItemResponse;
import org.elasticsearch.action.bulk.BulkRequestBuilder;
import org.elasticsearch.action.bulk.BulkResponse;
import org.elasticsearch.action.index.IndexRequest;
import org.elasticsearch.action.update.UpdateRequest;
import org.elasticsearch.client.Client;
import org.elasticsearch.client.Requests;
import org.elasticsearch.client.transport.TransportClient;
import org.elasticsearch.common.settings.Settings;
import org.elasticsearch.common.transport.InetSocketTransportAddress;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.apache.kafka.connect.es.KafkaElasticSearchSinkConnectorConfig.*;

import java.net.InetAddress;
import java.util.*;

/**
 * ElasticsearchSinkTask is a Task that takes records loaded from Kafka and sends them to
 * another system.
 */
public class KafkaElasticSearchSinkTask extends SinkTask {

    private static final Logger log = LoggerFactory.getLogger(KafkaElasticSearchSinkTask.class);

    private KafkaElasticSearchSinkConnectorConfig config;
    private Client client;
    private Map<String, String> topicIndexes;
    private Map<String, String> topicMappings;
    private ActionType actionType;
    private Integer bulkSize;
    private String idField;

    public KafkaElasticSearchSinkTask() {
        topicIndexes = new HashMap<>(0);
        topicMappings = new HashMap<>(0);
    }

    @Override
    public String version() {
        return new KafkaElasticSearchSinkConnector().version();
    }

    /**
     * Start the Task. Handles configuration parsing and one-time setup of the task.
     *
     * @param props initial configuration
     */
    @Override
    public void start(Map<String, String> props) {

        try {
            config = new KafkaElasticSearchSinkConnectorConfig(props);
        } catch (ConfigException e) {
            throw new ConnectException("Couldn't start " + KafkaElasticSearchSinkConnector.class.getName() + " due to configuration error.", e);
        }

        String clusterName = config.getString(KafkaElasticSearchSinkConnectorConfig.CLUSTER_NAME);
        String hosts = config.getString(KafkaElasticSearchSinkConnectorConfig.HOSTS);
        String topics = config.getString(KafkaElasticSearchSinkConnectorConfig.TOPICS);
        String indexes = config.getString(KafkaElasticSearchSinkConnectorConfig.INDEX);
        String mappingTypes = config.getString(KafkaElasticSearchSinkConnectorConfig.MAPPING_TYPE);
        actionType = config.getActionType(config.getString(KafkaElasticSearchSinkConnectorConfig.ACTION_TYPE));
        idField = config.getString(KafkaElasticSearchSinkConnectorConfig.ID_FIELD);
        bulkSize = config.getInt(KafkaElasticSearchSinkConnectorConfig.BULK_SIZE);

        List<String> hostsList = new ArrayList<>(Arrays.asList(hosts.replaceAll(" ", "").split(",")));
        List<String> topicsList = Arrays.asList(topics.replaceAll(" ", "").split(","));
        List<String> indexesList = Arrays.asList(indexes.replaceAll(" ", "").split(","));
        List<String> mappingTypesList = Arrays.asList(mappingTypes.replaceAll(" ", "").split(","));

        if (topicsList.size() != indexesList.size()) {
            throw new ConnectException("The number of indexes should be the same as the number of topics");
        }

        for (int i = 0; i < topicsList.size(); i++) {
            topicIndexes.put(topicsList.get(i), indexesList.get(i));
            topicMappings.put(topicsList.get(i), mappingTypesList.get(i));
        }

        try {
            Settings settings = Settings.settingsBuilder().put("cluster.name", clusterName).build();
            client = TransportClient.builder().settings(settings).build();
            for (String host : hostsList) {
                String address;
                Integer port;
                String[] hostArray = host.split(":");
                address = hostArray[0];

                try {
                    port = Integer.parseInt(hostArray[1]);
                } catch (Exception e) {
                    port = 9300;
                }
                log.info("address " + address + "port " + port);
                ((TransportClient) client).addTransportAddress(new InetSocketTransportAddress(InetAddress.getByName(address), port));
            }
        } catch (Exception e) {
            throw new ConnectException("Impossible to connect to hosts");
        }

    }

    /**
     * Put the records in the sink.
     *
     * @param sinkRecords the set of records to send.
     */
    @Override
    public void put(Collection<SinkRecord> sinkRecords) {
        try {
            List<SinkRecord> records = new ArrayList<>(sinkRecords);

            for (int i = 0; i < records.size(); i++) {
                BulkRequestBuilder bulkRequest = client.prepareBulk().setRefresh(Boolean.TRUE);

                for (int j = 0; j < bulkSize && i < records.size(); j++, i++) {
                    SinkRecord record = records.get(i);
                    Map<String, Object> jsonMap = (Map<String, Object>) record.value();

                    if (!jsonMap.isEmpty()) {
                        String index = topicIndexes.get(record.topic());
                        String mappingType = topicMappings.get(record.topic());
                        String id;

                        switch (actionType) {
                            case INSERT:
                                bulkRequest.add(client.prepareIndex(index, mappingType).setSource(jsonMap));
                                break;
                            case DELETE:
                                id = (String) jsonMap.get(idField);
                                bulkRequest.add(Requests.deleteRequest(index).type(mappingType).id(id));
                                break;
                            case UPDATE:
                                id = (String) jsonMap.get(idField);
                                bulkRequest.add(new UpdateRequest(index, mappingType, id).doc(jsonMap));
                                break;
                            case UPSERT:
                                id = (String) jsonMap.get(idField);
                                IndexRequest indexRequest = new IndexRequest(index, mappingType, id).source(jsonMap);
                                bulkRequest.add(new UpdateRequest(index, mappingType, id).doc(jsonMap)
                                        .upsert(indexRequest));
                                break;
                        }
                    }
                }
                i--;
                if (bulkRequest.numberOfActions() > 0) {
                    BulkResponse bulkResponse = bulkRequest.execute().actionGet();
                    if (bulkResponse.hasFailures()) {
                        for (BulkItemResponse item : bulkResponse) {
                            log.error(item.getFailureMessage());
                        }
                    }
                }
            }
        } catch (Exception e) {
            throw new RetriableException("Elasticsearch not connected", e);
        }
    }

    @Override
    public void flush(Map<TopicPartition, OffsetAndMetadata> offsets) {
    }

    @Override
    public void stop() {
        client.close();
    }

    public KafkaElasticSearchSinkConnectorConfig getConfig() {
        return config;
    }

    public void setConfig(KafkaElasticSearchSinkConnectorConfig config) {
        this.config = config;
    }

    public Client getClient() {
        return client;
    }

    public void setClient(Client client) {
        this.client = client;
    }

    public Map<String, String> getTopicIndexes() {
        return topicIndexes;
    }

    public void setTopicIndexes(Map<String, String> topicIndexes) {
        this.topicIndexes = topicIndexes;
    }

    public Map<String, String> getTopicMappings() {
        return topicMappings;
    }

    public void setTopicMappings(Map<String, String> topicMappings) {
        this.topicMappings = topicMappings;
    }

    public ActionType getActionType() {
        return actionType;
    }

    public void setActionType(ActionType actionType) {
        this.actionType = actionType;
    }

    public Integer getBulkSize() {
        return bulkSize;
    }

    public void setBulkSize(Integer bulkSize) {
        this.bulkSize = bulkSize;
    }

    public String getIdField() {
        return idField;
    }

    public void setIdField(String idField) {
        this.idField = idField;
    }
}

src/main/java/org/apache/kafka/connect/es/KafkaElasticSearchSinkConnector.java
/*
 * Copyright (C) 2015 Stratio (http://stratio.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.kafka.connect.es;

import org.apache.kafka.common.config.ConfigDef;
import org.apache.kafka.common.utils.AppInfoParser;
import org.apache.kafka.connect.connector.Task;
import org.apache.kafka.connect.errors.ConnectException;
import org.apache.kafka.connect.sink.SinkConnector;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;


/**
 * HdfsSinkConnector is a Kafka Connect Connector implementation that ingest data from Kafka to HDFS.
 */
public class KafkaElasticSearchSinkConnector extends SinkConnector {

    Map<String, String> configProperties;
    KafkaElasticSearchSinkConnectorConfig config;

    @Override
    public String version() {
        return AppInfoParser.getVersion();
    }

    @Override
    public void start(Map<String, String> props) throws ConnectException {
        try {
            configProperties = props;
            config = new KafkaElasticSearchSinkConnectorConfig(props);
            //} catch (ConfigException e) {
        } catch (Exception e) {
            throw new ConnectException("Couldn't start KafkaElasticSearchSinkConnector due to configuration error", e);
        }
    }

    @Override
    public Class<? extends Task> taskClass() {
        return KafkaElasticSearchSinkTask.class;
    }

    @Override
    public List<Map<String, String>> taskConfigs(int maxTasks) {
        List<Map<String, String>> taskConfigs = new ArrayList<>();
        Map<String, String> taskProps = new HashMap<>();
        taskProps.putAll(configProperties);
        for (int i = 0; i < maxTasks; i++) {
            taskConfigs.add(taskProps);
        }
        return taskConfigs;
    }

    @Override
    public void stop() throws ConnectException {

    }

    @Override
    public ConfigDef config() {
        return KafkaElasticSearchSinkConnectorConfig.config;
    }
}

src/main/java/org/apache/kafka/connect/es/KafkaElasticSearchSinkConnectorConfig.java
/*
 * Copyright (C) 2015 Stratio (http://stratio.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.kafka.connect.es;

import java.util.Map;

import org.apache.kafka.common.config.AbstractConfig;
import org.apache.kafka.common.config.ConfigDef;
import org.apache.kafka.common.config.ConfigDef.Importance;
import org.apache.kafka.common.config.ConfigDef.Type;

import java.util.HashMap;


public class KafkaElasticSearchSinkConnectorConfig extends AbstractConfig {

    static final String CLUSTER_NAME = "elasticsearch.cluster.name";
    static final String HOSTS = "elasticsearch.hosts";
    static final String BULK_SIZE = "elasticsearch.bulk.size";
    static final String MAPPING_TYPE = "elasticsearch.mapping.type";
    static final String TOPICS = "topics";
    static final String INDEX = "elasticsearch.index";
    static final String ID_FIELD = "elasticsearch.idField";
    static final String ACTION_TYPE = "action.type";

    private static final String DEFAULT_ACTION_TYPE = "insert";
    private static final String DEFAULT_ID_FIELD = "id";
    private static final Integer DEFAULT_BULK_SIZE = 250;

    private static final String CLUSTER_NAME_DOC = "Elastic search cluster name";
    private static final String HOSTS_DOC = "A comma separated Elastic search hosts including port with a :. For " +
            "example localhost:9300";
    private static final String BULK_SIZE_DOC = "The number of messages to be bulk indexed into elasticsearch";
    private static final String MAPPING_TYPE_DOC = "The mapping associated to the topic-indexes";
    private static final String TOPICS_DOC = "topics in kafka";
    private static final String INDEX_DOC = "The name of elasticsearch index";
    private static final String ID_FIELD_DOC = "Id field when the action type is delete, update or upsert";
    private static final String ACTION_TYPE_DOC = "The action type against how the messages should be processed." +
            " Default is: index. The following options are available:\n"
            + "insert : Creates documents in ES with the values in the received message\n"
            + "update : Update documents in ES with the values in the received message based on id field\n" +
            "upsert : Update and create if not exists documents in ES with the values in the received " +
            "message based on id field set\n"
            + "delete : Deletes documents from ES based on id field in the received message";

    static ConfigDef config = new ConfigDef().define(HOSTS, Type.STRING, Importance.HIGH, HOSTS_DOC)
            .define(CLUSTER_NAME, Type.STRING, Importance.HIGH, CLUSTER_NAME_DOC)
            .define(INDEX, Type.STRING, Importance.HIGH, INDEX_DOC)
            .define(BULK_SIZE, Type.INT, DEFAULT_BULK_SIZE, Importance.HIGH, BULK_SIZE_DOC)
            .define(TOPICS, Type.STRING, Importance.HIGH, TOPICS_DOC)
            .define(MAPPING_TYPE, Type.STRING, Importance.HIGH, MAPPING_TYPE_DOC)
            .define(ID_FIELD, Type.STRING, DEFAULT_ID_FIELD, Importance.MEDIUM, ID_FIELD_DOC)
            .define(ACTION_TYPE, Type.STRING, DEFAULT_ACTION_TYPE, Importance.HIGH, ACTION_TYPE_DOC);

    KafkaElasticSearchSinkConnectorConfig(Map<String, String> props) {
        super(config, props);
    }

    enum ActionType {

        DELETE("delete"), UPDATE("update"), UPSERT("upsert"), INSERT("insert");

        private String actionType;
        private static Map<String, ActionType> ACTIONS = init();

        ActionType(String actionType) {
            this.actionType = actionType;
        }

        public String toValue() {
            return actionType;
        }

        public static ActionType toType(String value) {
            return ACTIONS.get(value);
        }

        public static Map<String, ActionType> init() {
            Map<String, ActionType> actions = new HashMap<>();
            ActionType[] types = values();
            for (ActionType type : types) {
                actions.put(type.name().toLowerCase(), type);
            }
            return actions;
        }
    }

    ActionType getActionType(String actionType) {
        return ActionType.toType(actionType);
    }

}